Azure Blob Storage: Shared access signature (SAS) authorisation

Hi all,

I am intending to use Great Expectations on data that is stored in Azure Blob Storage (ASB) container using SAS token instead of access key. 

I have followed these instructions: https://docs.greatexpectations.io/docs/guides/connecting_to_your_data/cloud/azure/spark

What I have established is that when account_url/account_key pair is specified in data_connectors part of the *.yaml configuration file (see below), verification of the new datasource as described in [step 5 of the tutorial](https://docs.greatexpectations.io/docs/guides/connecting_to_your_data/cloud/azure/spark#5-test-your-new-datasource) succeeds. 

However, when account_url/SAS token pair is provided, I get:

 `java.lang.IllegalArgumentException: The String is not a valid Base64-encoded string.` error.

```
example_yaml = rf"""
name: my_azure_datasource
class_name: Datasource
execution_engine:
    class_name: SparkDFExecutionEngine
    azure_options:
        account_url: <mystorageaccount>.blob.core.windows.net
        credential: <SAS_TOKEN_OR_ACCOUNT_KEY>
data_connectors:
    default_runtime_data_connector_name:
        class_name: RuntimeDataConnector
        batch_identifiers:
            - default_identifier_name
    default_inferred_data_connector_name:
        class_name: InferredAssetAzureDataConnector
        azure_options:
            account_url: <mystorageaccount>.blob.core.windows.net
            credential: <SAS_TOKEN_OR_ACCOUNT_KEY>
        container: my-container
        name_starts_with: my_data
        default_regex:
            pattern: (.*)\.parquet
            group_names:
                - data_asset_name
"""
```

Furthermore, continuing with account_url/account_key pair, I tried to create an expectation suite using following commands:


```
**~$ great_expectations suite new**
Using v3 (Batch Request) API
Using Spark's default log4j profile: org/apache/spark/log4j-defaults.properties
Setting default log level to "WARN".
To adjust logging level use sc.setLogLevel(newLevel). For SparkR, use setLogLevel(newLevel).
22/04/21 16:36:38 WARN NativeCodeLoader: Unable to load native-hadoop library for your platform... using builtin-java classes where applicable
22/04/21 16:36:39 WARN Utils: Service 'SparkUI' could not bind on port 4040. Attempting port 4041.
22/04/21 16:36:39 WARN Utils: Service 'SparkUI' could not bind on port 4041. Attempting port 4042.

**How would you like to create your Expectation Suite?**
    1. Manually, without interacting with a sample batch of data (default)
    2. Interactively, with a sample batch of data
    3. Automatically, using a profiler
: 3

**A batch of data is required to edit the suite - let's help you to specify it.**

**Which data asset (accessible by data connector "default_inferred_data_connector_name") would you like to use?**
    1. somefile

Type [n] to see the next page or [p] for the previous. When you're ready to select an asset, enter the index.
: 1

Name the new Expectation Suite [somefile]: conf_test

Great Expectations will create a notebook, containing code cells that select from available columns in your dataset and
generate expectations about them to demonstrate some examples of assertions you can make about your data.

When you run this notebook, Great Expectations will store these expectations in a new Expectation Suite "exp_test" here:

  file:///home/<USER_NAME>/great_expectations/expectations/exp_test.json

Would you like to proceed? [Y/n]: Y
22/04/21 16:39:22 WARN FileStreamSink: Assume no metadata directory. Error while looking for metadata directory in the path: wasbs://my-file-system@<mystorageaccount>.blob.core.windows.net/somefile.csv.
java.lang.RuntimeException: java.lang.ClassNotFoundException: Class org.apache.hadoop.fs.azure.NativeAzureFileSystem$Secure not found
        at org.apache.hadoop.conf.Configuration.getClass(Configuration.java:2667)
        at org.apache.hadoop.fs.FileSystem.getFileSystemClass(FileSystem.java:3431)
        at org.apache.hadoop.fs.FileSystem.createFileSystem(FileSystem.java:3466)
        at org.apache.hadoop.fs.FileSystem.access$300(FileSystem.java:174)
        at org.apache.hadoop.fs.FileSystem$Cache.getInternal(FileSystem.java:3574)
        at org.apache.hadoop.fs.FileSystem$Cache.get(FileSystem.java:3521)
        at org.apache.hadoop.fs.FileSystem.get(FileSystem.java:540)
        at org.apache.hadoop.fs.Path.getFileSystem(Path.java:365)
        at org.apache.spark.sql.execution.streaming.FileStreamSink$.hasMetadata(FileStreamSink.scala:53)
        at org.apache.spark.sql.execution.datasources.DataSource.resolveRelation(DataSource.scala:370)
        at org.apache.spark.sql.DataFrameReader.loadV1Source(DataFrameReader.scala:274)
        at org.apache.spark.sql.DataFrameReader.$anonfun$load$3(DataFrameReader.scala:245)
        at scala.Option.getOrElse(Option.scala:189)
        at org.apache.spark.sql.DataFrameReader.load(DataFrameReader.scala:245)
        at org.apache.spark.sql.DataFrameReader.csv(DataFrameReader.scala:571)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at py4j.reflection.MethodInvoker.invoke(MethodInvoker.java:244)
        at py4j.reflection.ReflectionEngine.invoke(ReflectionEngine.java:357)
        at py4j.Gateway.invoke(Gateway.java:282)
        at py4j.commands.AbstractCommand.invokeMethod(AbstractCommand.java:132)
        at py4j.commands.CallCommand.execute(CallCommand.java:79)
        at py4j.ClientServerConnection.waitForCommands(ClientServerConnection.java:182)
        at py4j.ClientServerConnection.run(ClientServerConnection.java:106)
        at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.ClassNotFoundException: Class org.apache.hadoop.fs.azure.NativeAzureFileSystem$Secure not found
        at org.apache.hadoop.conf.Configuration.getClassByName(Configuration.java:2571)
        at org.apache.hadoop.conf.Configuration.getClass(Configuration.java:2665)
        ... 26 more
Traceback (most recent call last):
  File "/opt/conda/bin/great_expectations", line 10, in <module>
    sys.exit(main())
  File "/opt/conda/lib/python3.7/site-packages/great_expectations/cli/cli.py", line 184, in main
    cli()
  File "/opt/conda/lib/python3.7/site-packages/click/core.py", line 829, in __call__
    return self.main(*args, **kwargs)
  File "/opt/conda/lib/python3.7/site-packages/click/core.py", line 782, in main
    rv = self.invoke(ctx)
  File "/opt/conda/lib/python3.7/site-packages/click/core.py", line 1259, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/opt/conda/lib/python3.7/site-packages/click/core.py", line 1259, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/opt/conda/lib/python3.7/site-packages/click/core.py", line 1066, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/opt/conda/lib/python3.7/site-packages/click/core.py", line 610, in invoke
    return callback(*args, **kwargs)
  File "/opt/conda/lib/python3.7/site-packages/click/decorators.py", line 21, in new_func
    return f(get_current_context(), *args, **kwargs)
  File "/opt/conda/lib/python3.7/site-packages/great_expectations/cli/suite.py", line 143, in suite_new
    batch_request=batch_request,
  File "/opt/conda/lib/python3.7/site-packages/great_expectations/cli/suite.py", line 318, in _suite_new_workflow
    raise e
  File "/opt/conda/lib/python3.7/site-packages/great_expectations/cli/suite.py", line 294, in _suite_new_workflow
    assume_yes=False,
  File "/opt/conda/lib/python3.7/site-packages/great_expectations/cli/suite.py", line 844, in _suite_edit_workflow
    raise e
  File "/opt/conda/lib/python3.7/site-packages/great_expectations/cli/suite.py", line 781, in _suite_edit_workflow
    batch_request=batch_request,
  File "/opt/conda/lib/python3.7/site-packages/great_expectations/render/renderer/v3/suite_profile_notebook_renderer.py", line 40, in __init__
    expectation_suite_name=expectation_suite_name,
  File "/opt/conda/lib/python3.7/site-packages/great_expectations/data_context/data_context.py", line 1862, in get_validator
    **kwargs,
  File "/opt/conda/lib/python3.7/site-packages/great_expectations/core/usage_statistics/usage_statistics.py", line 287, in usage_statistics_wrapped_method
    result = func(*args, **kwargs)
  File "/opt/conda/lib/python3.7/site-packages/great_expectations/data_context/data_context.py", line 1749, in get_batch_list
    return datasource.get_batch_list_from_batch_request(batch_request=batch_request)
  File "/opt/conda/lib/python3.7/site-packages/great_expectations/datasource/new_datasource.py", line 195, in get_batch_list_from_batch_request
    batch_definition=batch_definition
  File "/opt/conda/lib/python3.7/site-packages/great_expectations/datasource/data_connector/data_connector.py", line 111, in get_batch_data_and_metadata
    batch_spec=batch_spec
  File "/opt/conda/lib/python3.7/site-packages/great_expectations/execution_engine/sparkdf_execution_engine.py", line 264, in get_batch_data_and_markers
    batch_data = reader_fn(path)
  File "/opt/conda/lib/python3.7/site-packages/pyspark/sql/readwriter.py", line 410, in csv
    return self._df(self._jreader.csv(self._spark._sc._jvm.PythonUtils.toSeq(path)))
  File "/opt/conda/lib/python3.7/site-packages/py4j/java_gateway.py", line 1322, in __call__
    answer, self.gateway_client, self.target_id, self.name)
  File "/opt/conda/lib/python3.7/site-packages/pyspark/sql/utils.py", line 111, in deco
    return f(*a, **kw)
  File "/opt/conda/lib/python3.7/site-packages/py4j/protocol.py", line 328, in get_return_value
    format(target_id, ".", name), value)
py4j.protocol.Py4JJavaError: An error occurred while calling o113.csv.
: java.lang.RuntimeException: java.lang.ClassNotFoundException: Class org.apache.hadoop.fs.azure.NativeAzureFileSystem$Secure not found
        at org.apache.hadoop.conf.Configuration.getClass(Configuration.java:2667)
        at org.apache.hadoop.fs.FileSystem.getFileSystemClass(FileSystem.java:3431)
        at org.apache.hadoop.fs.FileSystem.createFileSystem(FileSystem.java:3466)
        at org.apache.hadoop.fs.FileSystem.access$300(FileSystem.java:174)
        at org.apache.hadoop.fs.FileSystem$Cache.getInternal(FileSystem.java:3574)
        at org.apache.hadoop.fs.FileSystem$Cache.get(FileSystem.java:3521)
        at org.apache.hadoop.fs.FileSystem.get(FileSystem.java:540)
        at org.apache.hadoop.fs.Path.getFileSystem(Path.java:365)
        at org.apache.spark.sql.execution.datasources.DataSource$.$anonfun$checkAndGlobPathIfNecessary$1(DataSource.scala:747)
        at scala.collection.immutable.List.map(List.scala:293)
        at org.apache.spark.sql.execution.datasources.DataSource$.checkAndGlobPathIfNecessary(DataSource.scala:745)
        at org.apache.spark.sql.execution.datasources.DataSource.checkAndGlobPathIfNecessary(DataSource.scala:577)
        at org.apache.spark.sql.execution.datasources.DataSource.resolveRelation(DataSource.scala:408)
        at org.apache.spark.sql.DataFrameReader.loadV1Source(DataFrameReader.scala:274)
        at org.apache.spark.sql.DataFrameReader.$anonfun$load$3(DataFrameReader.scala:245)
        at scala.Option.getOrElse(Option.scala:189)
        at org.apache.spark.sql.DataFrameReader.load(DataFrameReader.scala:245)
        at org.apache.spark.sql.DataFrameReader.csv(DataFrameReader.scala:571)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at py4j.reflection.MethodInvoker.invoke(MethodInvoker.java:244)
        at py4j.reflection.ReflectionEngine.invoke(ReflectionEngine.java:357)
        at py4j.Gateway.invoke(Gateway.java:282)
        at py4j.commands.AbstractCommand.invokeMethod(AbstractCommand.java:132)
        at py4j.commands.CallCommand.execute(CallCommand.java:79)
        at py4j.ClientServerConnection.waitForCommands(ClientServerConnection.java:182)
        at py4j.ClientServerConnection.run(ClientServerConnection.java:106)
        at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.ClassNotFoundException: Class org.apache.hadoop.fs.azure.NativeAzureFileSystem$Secure not found
        at org.apache.hadoop.conf.Configuration.getClassByName(Configuration.java:2571)
        at org.apache.hadoop.conf.Configuration.getClass(Configuration.java:2665)
        ... 29 more
```

Hence, I have two questions:

1. Is it possible to use SAS tokens in order to access data stored in ABS container or only account_url/account_key pair is allowed?

2. Given successful configuration of the data connector using account_url/account_key pair, what might cause problems during suite creation?

Kind regards

--- a/docs/guides/setup/configuring_data_docs/how_to_host_and_share_data_docs_on_azure_blob_storage.md
+++ b/docs/guides/setup/configuring_data_docs/how_to_host_and_share_data_docs_on_azure_blob_storage.md
@@ -56,7 +56,7 @@ data_docs_sites:
     store_backend:
        class_name: TupleAzureBlobStoreBackend
        container: \$web
-       connection_string: ${AZURE_STORAGE_WEB_CONNECTION_STRING}
+       connection_string: ${AZURE_STORAGE_CONNECTION_STRING}
     site_index_builder:
       class_name: DefaultSiteIndexBuilder
 ```
@@ -64,12 +64,30 @@ data_docs_sites:
 You may also replace the default ``local_site`` if you would only like to maintain a single Azure Data Docs site.
 
 :::note
- Since the container is called ``$web``, if we simply set ``container: $web`` in ``great_expectations.yml`` then Great Expectations would unsuccefully try to find the variable called ``web`` in ``config_variables.yml``. 
+ Since the container is called ``$web``, if we simply set ``container: $web`` in ``great_expectations.yml`` then Great Expectations would unsuccessfully try to find the variable called ``web`` in ``config_variables.yml``. 
  We use an escape char ``\`` before the ``$`` so the [substitute_config_variable](https://legacy.docs.greatexpectations.io/en/latest/autoapi/great_expectations/data_context/util/index.html?highlight=substitute_config_variable#great_expectations.data_context.util.substitute_config_variable) method will allow us to reach the ``$web`` container.
 :::
 
 You also may configure Great Expectations to store your <TechnicalTag relative="../../../" tag="expectation" text="Expectations" /> and <TechnicalTag relative="../../../" tag="validation_result" text="Validation Results" /> in this Azure Storage account.
 You can follow the documentation from the guides for [Expectations](../../setup/configuring_metadata_stores/how_to_configure_an_expectation_store_in_azure_blob_storage.md) and [Validation Results](../../setup/configuring_metadata_stores/how_to_configure_a_validation_result_store_in_azure_blob_storage.md) but be sure you set ``container: \$web`` in place of the other container name.
+
+The following options are available for this backend:
+
+  * ``container``: The name of the Azure Blob container to store your data in.
+  * ``connection_string``: The Azure Storage connection string.
+  This can also be supplied by setting the ``AZURE_STORAGE_CONNECTION_STRING`` environment variable.
+  * ``prefix``: All paths on blob storage will be prefixed with this string.
+  * ``account_url``: The URL to the blob storage account. Any other entities included in the URL path (e.g. container or blob) will be discarded.
+  This URL can be optionally authenticated with a SAS token.
+  This can only be used if you don't configure the ``connection_string``. 
+  You can also configure this by setting the ``AZURE_STORAGE_ACCOUNT_URL`` environment variable.
+
+The most common authentication methods are supported:
+
+* SAS token authentication: append the SAS token to ``account_url`` or make sure it is set in the ``connection_string``.
+* Account key authentication: include the account key in the ``connection_string``.
+* When none of the above authentication methods are specified, the [DefaultAzureCredential](https://docs.microsoft.com/en-us/python/api/azure-identity/azure.identity.defaultazurecredential?view=azure-python) will be used which supports most common authentication methods.
+  You still need to provide the account url either through the config file or environment variable.
 
 
 ### 4. Build the Azure Blob Data Docs site

--- a/docs/guides/setup/configuring_metadata_stores/how_to_configure_a_validation_result_store_in_azure_blob_storage.md
+++ b/docs/guides/setup/configuring_metadata_stores/how_to_configure_a_validation_result_store_in_azure_blob_storage.md
@@ -65,6 +65,12 @@ stores:
 If the container is called ``$web`` (for [hosting and sharing Data Docs on Azure Blob Storage](../../setup/configuring_data_docs/how_to_host_and_share_data_docs_on_azure_blob_storage.md)) then set ``container: \$web`` so the escape char will allow us to reach the ``$web``container.
 :::
 
+:::note
+Various authentication and configuration options are available as documented in [hosting and sharing Data Docs on Azure Blob Storage](../../setup/configuring_data_docs/how_to_host_and_share_data_docs_on_azure_blob_storage.md).
+:::
+
+
+
 ### 4. Copy existing Validation Results JSON files to the Azure blob (This step is optional)
 
 One way to copy Validation Results into Azure Blob Storage is by using the ``az storage blob upload`` command, which is part of the Azure SDK. The following example will copy one Validation Result from a local folder to the Azure blob.   Information on other ways to copy Validation Result JSON files, like the Azure Storage browser in the Azure Portal, can be found in the [Documentation for Azure](https://docs.microsoft.com/en-us/azure/storage/blobs/storage-quickstart-blobs-portal).

--- a/docs/guides/setup/configuring_metadata_stores/how_to_configure_an_expectation_store_in_azure_blob_storage.md
+++ b/docs/guides/setup/configuring_metadata_stores/how_to_configure_an_expectation_store_in_azure_blob_storage.md
@@ -65,6 +65,10 @@ stores:
 If the container is called ``$web`` (for [hosting and sharing Data Docs on Azure Blob Storage](../configuring_data_docs/how_to_host_and_share_data_docs_on_azure_blob_storage.md)) then set ``container: \$web`` so the escape char will allow us to reach the ``$web``container.
 :::
 
+:::note
+Various authentication and configuration options are available as documented in [hosting and sharing Data Docs on Azure Blob Storage](../../setup/configuring_data_docs/how_to_host_and_share_data_docs_on_azure_blob_storage.md).
+:::
+
 
 ### 4. Copy existing Expectation JSON files to the Azure blob (This step is optional)
 

--- a/great_expectations/data_context/store/tuple_store_backend.py
+++ b/great_expectations/data_context/store/tuple_store_backend.py
@@ -970,8 +970,9 @@ class TupleAzureBlobStoreBackend(TupleStoreBackend):
     def __init__(
         self,
         container,
-        connection_string,
-        prefix="",
+        connection_string=None,
+        account_url=None,
+        prefix=None,
         filepath_template=None,
         filepath_prefix=None,
         filepath_suffix=None,
@@ -993,9 +994,12 @@ class TupleAzureBlobStoreBackend(TupleStoreBackend):
             manually_initialize_store_backend_id=manually_initialize_store_backend_id,
             store_name=store_name,
         )
-        self.connection_string = connection_string
-        self.prefix = prefix
+        self.connection_string = connection_string or os.environ.get(
+            "AZURE_STORAGE_CONNECTION_STRING"
+        )
+        self.prefix = prefix or ""
         self.container = container
+        self.account_url = account_url or os.environ.get("AZURE_STORAGE_ACCOUNT_URL")
 
     def _get_container_client(self):
 
@@ -1004,6 +1008,10 @@ class TupleAzureBlobStoreBackend(TupleStoreBackend):
         if self.connection_string:
             return BlobServiceClient.from_connection_string(
                 self.connection_string
+            ).get_container_client(self.container)
+        elif self.account_url:
+            return BlobServiceClient(
+                account_url=self.account_url, credential=DefaultAzureCredential()
             ).get_container_client(self.container)
         else:
             raise StoreBackendError(

--- a/great_expectations/data_context/store/tuple_store_backend.py
+++ b/great_expectations/data_context/store/tuple_store_backend.py
@@ -1003,7 +1003,7 @@ class TupleAzureBlobStoreBackend(TupleStoreBackend):
         self.account_url = account_url or os.environ.get("AZURE_STORAGE_ACCOUNT_URL")
 
     @property
-    @functools.lru_cache
+    @functools.lru_cache()
     def _container_client(self):
 
         from azure.identity import DefaultAzureCredential

--- a/great_expectations/data_context/store/tuple_store_backend.py
+++ b/great_expectations/data_context/store/tuple_store_backend.py
@@ -1,11 +1,11 @@
 # PYTHON 2 - py2 - update to ABC direct use rather than __metaclass__ once we drop py2 support
+import functools
 import logging
 import os
 import random
 import re
 import shutil
 from abc import ABCMeta
-from functools import cached_property
 from typing import List, Tuple
 
 from great_expectations.data_context.store.store_backend import StoreBackend
@@ -1002,7 +1002,8 @@ class TupleAzureBlobStoreBackend(TupleStoreBackend):
         self.container = container
         self.account_url = account_url or os.environ.get("AZURE_STORAGE_ACCOUNT_URL")
 
-    @cached_property
+    @property
+    @functools.lru_cache
     def _container_client(self):
 
         from azure.identity import DefaultAzureCredential

--- a/great_expectations/data_context/store/tuple_store_backend.py
+++ b/great_expectations/data_context/store/tuple_store_backend.py
@@ -5,6 +5,7 @@ import random
 import re
 import shutil
 from abc import ABCMeta
+from functools import cached_property
 from typing import List, Tuple
 
 from great_expectations.data_context.store.store_backend import StoreBackend
@@ -1001,30 +1002,31 @@ class TupleAzureBlobStoreBackend(TupleStoreBackend):
         self.container = container
         self.account_url = account_url or os.environ.get("AZURE_STORAGE_ACCOUNT_URL")
 
-    def _get_container_client(self):
+    @cached_property
+    def _container_client(self):
 
+        from azure.identity import DefaultAzureCredential
         from azure.storage.blob import BlobServiceClient
 
         if self.connection_string:
-            return BlobServiceClient.from_connection_string(
+            blob_service_client = BlobServiceClient.from_connection_string(
                 self.connection_string
-            ).get_container_client(self.container)
+            )
         elif self.account_url:
-            return BlobServiceClient(
+            blob_service_client = BlobServiceClient(
                 account_url=self.account_url, credential=DefaultAzureCredential()
-            ).get_container_client(self.container)
+            )
         else:
             raise StoreBackendError(
                 "Unable to initialize ServiceClient, AZURE_STORAGE_CONNECTION_STRING should be set"
             )
 
+        return blob_service_client.get_container_client(self.container)
+
     def _get(self, key):
         az_blob_key = os.path.join(self.prefix, self._convert_key_to_filepath(key))
         return (
-            self._get_container_client()
-            .download_blob(az_blob_key)
-            .readall()
-            .decode("utf-8")
+            self._container_client.download_blob(az_blob_key).readall().decode("utf-8")
         )
 
     def _set(self, key, value, content_encoding="utf-8", **kwargs):
@@ -1036,7 +1038,7 @@ class TupleAzureBlobStoreBackend(TupleStoreBackend):
         if isinstance(value, str):
             if az_blob_key.endswith(".html"):
                 my_content_settings = ContentSettings(content_type="text/html")
-                self._get_container_client().upload_blob(
+                self._container_client.upload_blob(
                     name=az_blob_key,
                     data=value,
                     encoding=content_encoding,
@@ -1044,14 +1046,14 @@ class TupleAzureBlobStoreBackend(TupleStoreBackend):
                     content_settings=my_content_settings,
                 )
             else:
-                self._get_container_client().upload_blob(
+                self._container_client.upload_blob(
                     name=az_blob_key,
                     data=value,
                     encoding=content_encoding,
                     overwrite=True,
                 )
         else:
-            self._get_container_client().upload_blob(
+            self._container_client.upload_blob(
                 name=az_blob_key, data=value, overwrite=True
             )
         return az_blob_key
@@ -1060,9 +1062,7 @@ class TupleAzureBlobStoreBackend(TupleStoreBackend):
         # Note that the prefix arg is only included to maintain consistency with the parent class signature
         key_list = []
 
-        for obj in self._get_container_client().list_blobs(
-            name_starts_with=self.prefix
-        ):
+        for obj in self._container_client.list_blobs(name_starts_with=self.prefix):
             az_blob_key = os.path.relpath(obj.name)
             if az_blob_key.startswith(f"{self.prefix}/"):
                 az_blob_key = az_blob_key[len(self.prefix) + 1 :]
@@ -1084,7 +1084,7 @@ class TupleAzureBlobStoreBackend(TupleStoreBackend):
         az_blob_path = os.path.join(self.container, self.prefix, az_blob_key)
 
         return "https://{}.blob.core.windows.net/{}".format(
-            self._get_container_client().account_name,
+            self._container_client.account_name,
             az_blob_path,
         )
 
@@ -1101,8 +1101,8 @@ class TupleAzureBlobStoreBackend(TupleStoreBackend):
             dest_blob_path = os.path.join(self.prefix, dest_blob_path)
 
         # azure storage sdk does not have _move method
-        source_blob = self._get_container_client().get_blob_client(source_blob_path)
-        dest_blob = self._get_container_client().get_blob_client(dest_blob_path)
+        source_blob = self._container_client.get_blob_client(source_blob_path)
+        dest_blob = self._container_client.get_blob_client(dest_blob_path)
 
         dest_blob.start_copy_from_url(source_blob.url, requires_sync=True)
         copy_properties = dest_blob.get_blob_properties().copy
@@ -1122,7 +1122,7 @@ class TupleAzureBlobStoreBackend(TupleStoreBackend):
         if not az_blob_path.startswith(self.prefix):
             az_blob_path = os.path.join(self.prefix, az_blob_path)
 
-        blob = self._get_container_client().get_blob_client(az_blob_path)
+        blob = self._container_client.get_blob_client(az_blob_path)
         blob.delete_blob()
         return True
 

--- a/requirements-dev-azure.txt
+++ b/requirements-dev-azure.txt
@@ -1,3 +1,3 @@
-azure-identity>=1.0.0
+azure-identity>=1.10.0
 azure-keyvault-secrets>=4.0.0
 azure-storage-blob>=12.5.0

--- a/tests/data_context/store/test_store_backends.py
+++ b/tests/data_context/store/test_store_backends.py
@@ -2,12 +2,13 @@ import datetime
 import json
 import os
 from collections import OrderedDict
-from unittest.mock import patch
+from unittest.mock import MagicMock, Mock, patch
 
 import boto3
 import pyparsing as pp
 import pytest
 from moto import mock_s3
+from pytest_mock import MockerFixture
 
 import tests.test_utils as test_utils
 from great_expectations.core.data_context_key import DataContextVariableKey
@@ -1191,14 +1192,14 @@ def test_TupleGCSStoreBackend():
     )
 
 
-def test_TupleAzureBlobStoreBackend():
-    pytest.importorskip("azure-storage-blob")
+def test_TupleAzureBlobStoreBackend_connection_string():
+    pytest.importorskip("azure.storage.blob")
     """
     What does this test test and why?
     Since no package like moto exists for Azure-Blob services, we mock the Azure-blob client
     and assert that the store backend makes the right calls for set, get, and list.
     """
-    connection_string = "this_is_a_test_conn_string"
+    connection_string = "DefaultEndpointsProtocol=https;AccountName=dummy;AccountKey=secret;EndpointSuffix=core.windows.net"
     prefix = "this_is_a_test_prefix"
     container = "dummy-container"
 
@@ -1210,53 +1211,58 @@ def test_TupleAzureBlobStoreBackend():
         "azure.storage.blob.BlobServiceClient", autospec=True
     ) as mock_azure_blob_client:
 
-        mock_container_client = mock_azure_blob_client.get_container_client.return_value
+        mock_container_client = my_store._container_client
+        mock_azure_blob_client.from_connection_string.assert_called_once()
 
         my_store.set(("AAA",), "aaa")
-
-        mock_azure_blob_client.from_connection_string.assert_called_once()
-        mock_container_client.assert_called_once()
         mock_container_client.upload_blob.assert_called_once_with(
-            name="AAA", data=b"aaa", encoding="utf-8"
+            name="this_is_a_test_prefix/AAA",
+            data="aaa",
+            encoding="utf-8",
+            overwrite=True,
         )
-
-    with patch(
-        "azure.storage.blob.BlobServiceClient", autospec=True
-    ) as mock_azure_blob_client:
-
-        mock_container_client = mock_azure_blob_client.get_container_client.return_value
-
-        my_store.set(("BBB",), b"bbb")
-
-        mock_azure_blob_client.from_connection_string.assert_called_once()
-        mock_container_client.assert_called_once()
-        mock_container_client.upload_blob.assert_called_once_with(
-            name="AAA", data=b"aaa"
-        )
-
-    with patch(
-        "azure.storage.blob.BlobServiceClient", autospec=True
-    ) as mock_azure_blob_client:
-
-        mock_container_client = mock_azure_blob_client.get_container_client.return_value
 
         my_store.get(("BBB",))
+        mock_container_client.download_blob.assert_called_once_with(
+            "this_is_a_test_prefix/BBB"
+        )
 
-        mock_azure_blob_client.from_connection_string.assert_called_once()
-        mock_container_client.assert_called_once()
-        mock_container_client.download_blob.assert_called_once_with("BBB")
+        my_store.list_keys()
+        mock_container_client.list_blobs.assert_called_once_with(
+            name_starts_with="this_is_a_test_prefix"
+        )
+
+
+def test_TupleAzureBlobStoreBackend_account_url():
+    pytest.importorskip("azure.storage.blob")
+    pytest.importorskip("azure.identity")
+    """
+    What does this test test and why?
+    Since no package like moto exists for Azure-Blob services, we mock the Azure-blob client
+    and assert that the store backend makes the right calls for set, get, and list.
+    """
+    account_url = "this_is_a_test_account_url"
+    prefix = "this_is_a_test_prefix"
+    container = "dummy-container"
+
+    my_store = TupleAzureBlobStoreBackend(
+        account_url=account_url, prefix=prefix, container=container
+    )
 
     with patch(
         "azure.storage.blob.BlobServiceClient", autospec=True
     ) as mock_azure_blob_client:
+        with patch(
+            "azure.identity.DefaultAzureCredential", autospec=True
+        ) as mock_azure_credential:
+            mock_container_client = my_store._container_client
+            mock_azure_blob_client.assert_called_once()
 
-        mock_container_client = mock_azure_blob_client.get_container_client.return_value
-
-        my_store.list_keys()
-
-        mock_azure_blob_client.from_connection_string.assert_called_once()
-        mock_container_client.assert_called_once()
-        mock_container_client.list_blobs.assert_called_once_with("this_is_a_prefix")
+            my_store.get(("BBB",))
+            mock_container_client.download_blob.assert_called_once_with(
+                "this_is_a_test_prefix/BBB"
+            )
+            mock_azure_credential.assert_called_once()
 
 
 @mock_s3

--- a/tests/data_context/store/test_store_backends.py
+++ b/tests/data_context/store/test_store_backends.py
@@ -8,7 +8,6 @@ import boto3
 import pyparsing as pp
 import pytest
 from moto import mock_s3
-from pytest_mock import MockerFixture
 
 import tests.test_utils as test_utils
 from great_expectations.core.data_context_key import DataContextVariableKey

--- a/tests/data_context/store/test_store_backends.py
+++ b/tests/data_context/store/test_store_backends.py
@@ -1193,6 +1193,7 @@ def test_TupleGCSStoreBackend():
 
 def test_TupleAzureBlobStoreBackend_connection_string():
     pytest.importorskip("azure.storage.blob")
+    pytest.importorskip("azure.identity")
     """
     What does this test test and why?
     Since no package like moto exists for Azure-Blob services, we mock the Azure-blob client

